site/src/jsMain/kotlin/com/varabyte/kobweb/site/components/widgets/video/YouTubeVideo.kt

```
@@ -0,0 +1,89 @@
+import com.varabyte.kobweb.compose.ui.modifiers.fillMaxWidth
+import com.varabyte.kobweb.compose.ui.modifiers.width
+import com.varabyte.kobweb.silk.style.CssStyle
+import com.varabyte.kobweb.silk.style.breakpoint.Breakpoint
+import com.varabyte.kobweb.silk.style.toAttrs
+import com.varabyte.kobweb.silk.style.toModifier
+import org.jetbrains.compose.web.css.percent
+import org.jetbrains.compose.web.dom.Iframe
+import org.w3c.dom.HTMLAnchorElement
+import org.w3c.dom.asList
+
+val YouTubeVideoStyle = CssStyle {
+    base {
+        Modifier.fillMaxWidth()
+            .aspectRatio(width = 16, height = 9)
+    }
+    Breakpoint.ZERO {
+        Modifier.width(85.percent)
+    }
+    Breakpoint.SM {
+        Modifier.width(86.percent)
+    }
+    Breakpoint.MD {
+        Modifier.width(88.percent)
+    }
+    Breakpoint.LG {
+        Modifier.width(90.percent)
+    }
+    Breakpoint.XL {
+        Modifier.width(100.percent)
+    }
+}
+
+/**
+ * Youtube Video Player
+ */
+@Composable
+fun YouTubeVideo(url: String? = null) {
+    var videoId by remember { mutableStateOf(url?.let { extractVideoId(url) }) }
+
+    Box(
+        modifier = YouTubeVideoStyle.toModifier(),
+        contentAlignment = Alignment.CenterStart,
+        ref = ref { element ->
+            if (videoId != null) return@ref
+            try {
+                var currentElement = element.parentElement
+                while (currentElement != null && videoId == null) {
+                    val links = currentElement.querySelectorAll("a")
+                    links.asList().forEach { node ->
+                        if (node is HTMLAnchorElement && node.href.contains("youtube.com")) {
+                            extractVideoId(node.href)?.let {
+                                videoId = it
+                                return@forEach
+                            }
+                        }
+                    }
+                    currentElement = currentElement.parentElement
```

I'm a little nervous about going up and up and up. I was hoping to find the URL within the YouTubePlayer's current parent. Does that not work? If we can only keep it to one parent that would be great.

The higher up you go, the more of the DOM you search, and the more likely you'll pick up a totally incorrect video URL.

(Obviously it doesn't matter that much for this site becuase we only have a single video page and we know all cases work..... but general advice is to write code as constrained as possible so it doesn't do something unexpected on you in the future. Better to fail fast than write code that is super robust but hard to reason about)

---

I  will try to solve this problem.